### DISCUSSION TITLE
JSON RPC Server Thread Safety

### DISCUSSION THREAD

**Discussion by majorpain1588** [stephenberry/glaze#291] [Q&A] [answered]
I've been playing around with the JSON-RPC server and was wondering about its thread safety properties.  Is it safe to share a single server instance between multiple threads that are servicing requests from different clients?

It seems like it works fine in a little test application I've been playing with and looking through the code I don't see any obvious issues. I figured I'd ask in case there are subtle details I'm missing.  If it safe to use in this way then maybe a quick sentence added the JSON RPC doc page would help answer future people's questions.

**Comment by jbbjarnason** [comment, marked as answer]
When implemented I had not multithreaded in mind and therefore it does not use any mutex locks. So in case of json rpc method you would need to finish serving one call before other can occur. 
But the state of the json rpc objects are rarely changing so carefully it could be run in multithreaded context. But I am not sure if you could bind a callback function to the object which should run in different thread, which is probably what you would like?

**Reply by jbbjarnason** [reply]
fyi, here is an example using zeromq req rep for server https://github.com/Skaginn3x/framework/blob/main/libs/rpc/inc/public/tfc/rpc.hpp#L82

the zeromq req rep pattern does not support doing requests in parallel  though
